mnemosyne/batch_tool.py

```
@@ -0,0 +1,199 @@
+    failed_index: int | None = None
+    action: str | None = None
+
+    def __str__(self) -> str:
+        return self.message
+
+
+class BatchOperationError(RuntimeError):
+    pass
+
+
+def validate_batch_operations(operations: Any) -> list[dict[str, Any]]:
```

Ruff (0.15.20)

[warning] Too many branches (14 > 12)

(PLR0912)